consensus/istanbul/backend/handler.go

```diff
@@ -290,3 +298,177 @@ func (sb *Backend) UnregisterPeer(peer consensus.Peer, isProxiedPeer bool) {
 		}
 	}
 }
+
+// Handshake allows this node to identify itself to the peer as a validator and vice versa
+func (sb *Backend) Handshake(peer consensus.Peer) (bool, error) {
+	errCh := make(chan error, 4)
+	isValidatorCh := make(chan bool, 1)
+
+	go func() {
+		validatorProofMessage, err := sb.generateValidatorProofMessage(peer)
+		errCh <- err
+		if err != nil {
+			return
+		}
+		msgBytes, err := validatorProofMessage.Payload()
+		errCh <- err
+		if err != nil {
+			return
+		}
+		errCh <- peer.Send(istanbulValidatorProofMsg, msgBytes)
+	}()
+	go func() {
+		isValidator, err := sb.readValidatorProofMessage(peer)
+		errCh <- err
+		isValidatorCh <- isValidator
+	}()
+
+	timeout := time.NewTimer(handshakeTimeout)
+	defer timeout.Stop()
+	// We write to errCh four times unless if the timeout is triggered or an error isn't nil
+	for i := 0; i < 4; i++ {
+		select {
+		case err := <-errCh:
+			if err != nil {
+				return false, err
+			}
+		case <-timeout.C:
+			return false, p2p.DiscReadTimeout
+		}
+	}
+	return <-isValidatorCh, nil
+}
+
+// generateValidatorProofMessage will create a message that contains a proof
+// to the peer during the handshake that this node is a validator, which is just
+// a versioned enode message.
+// If this node is not a validator or this node does not believe the peer
+// is a validator, no proof is generated to hide the address of this node
+// and an empty message is created.
+// If this node is a proxy, it will use the most recent versioned enode message this
+// node has received from the proxied validator.
+func (sb *Backend) generateValidatorProofMessage(peer consensus.Peer) (*istanbul.Message, error) {
+	shouldSend, err := sb.shouldSendValidatorProof(peer)
+	if err != nil {
+		return nil, err
+	}
+	if shouldSend {
+		msg, err := sb.getSelfVersionedEnodeMsg(getCurrentAnnounceVersion())
+		if err != nil {
+			return nil, err
+		}
+		if msg != nil {
+			return msg, nil
+		}
+	}
+	// Even if we decide not to identify ourselves,
+	// send an empty message to complete the handshake
+	return &istanbul.Message{}, nil
```

Comment:
Why can't we skip that part of the handshake? Based on my reading of the code above, this formulation adds at least one full RTT to the connection process (or more, see blow), even between non-validating fullnodes (but maybe I missed something). Is it possible to avoid this?

Reply:
The issue is that the eth handler that calls this handshake will wait for the handshake to be completed before continuing on with the rest of the setup (like registering the peer with `eth` and `istanbul` as ready for use). If I'm a node and my new peer doesn't send any message at all then I will end up having to wait the 5 second timeout to wait for a response. The 5 second wait is undesirable & currently the peer gets dropped if the handshake fails with an error (which timing out will result in)